Sync/Synchronizers/LoginsSynchronizer.swift

```diff
@@ -107,20 +108,20 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
         }
     }
 
-    private func uploadModifiedLogins(logins: [Login], lastTimestamp: Timestamp, fromStorage storage: SyncableLogins, withServer storageClient: Sync15CollectionClient<LoginPayload>) -> DeferredTimestamp {
-        return self.uploadRecords(logins.map(makeLoginRecord), by: 50, lastTimestamp: lastTimestamp, storageClient: storageClient) {
-            storage.markAsSynchronized($0.success, modified: $0.modified)
-        }
-    }
-
-    private func uploadDeletedLogins(guids: [GUID], lastTimestamp: Timestamp, fromStorage storage: SyncableLogins, withServer storageClient: Sync15CollectionClient<LoginPayload>) -> DeferredTimestamp {
-
-        let records = guids.map(makeDeletedLoginRecord)
+    private func uploadChangedRecords<T>(deleted: Set<GUID>, modified: Set<GUID>, records: [Record<T>], lastTimestamp: Timestamp,
+                                      storage: SyncableLogins, withServer storageClient: Sync15CollectionClient<T>) -> Success {
 
-        // Deletions are smaller, so upload 100 at a time.
-        return self.uploadRecords(records, by: 100, lastTimestamp: lastTimestamp, storageClient: storageClient) {
-            storage.markAsDeleted($0.success) >>> always($0.modified)
+        let onUpload: (POSTResult, Timestamp?) -> DeferredTimestamp = { result, lastModified in
+            let uploaded = Set(result.success)
+            let deletedGUIDs = Array(uploaded.intersect(deleted))
```

Comment:
Can you just make `markAsDeleted` and `markAsSynchronized` take a `SequenceType` or other abstraction and avoid the `Array` alloc? All it needs is to be able to walk the sequence…